File: test/fetch_test.go
package test

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/comoland/como/core"
	"github.com/comoland/como/js"
)

func TestFetch(t *testing.T) {
	// Create a test server
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/test":
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]string{"message": "test response"})
		case "/redirect":
			http.Redirect(w, r, "/test", http.StatusFound)
		case "/error":
			w.WriteHeader(http.StatusInternalServerError)
			w.Write([]byte("error response"))
		case "/timeout":
			time.Sleep(2 * time.Second)
			w.Write([]byte("timeout response"))
		default:
			w.WriteHeader(http.StatusNotFound)
		}
	}))
	defer ts.Close()

	// Create a new context
	Loop, ctx := core.Como("")
	defer ctx.Free()

	// Test cases
	tests := []struct {
		name     string
		url      string
		options  map[string]interface{}
		expected interface{}
	}{
		{
			name: "GET request",
			url:  ts.URL + "/test",
			options: map[string]interface{}{
				"method": "GET",
			},
			expected: map[string]string{"message": "test response"},
		},
		{
			name: "POST request",
			url:  ts.URL + "/test",
			options: map[string]interface{}{
				"method": "POST",
				"body":   "test body",
			},
			expected: map[string]string{"message": "test response"},
		},
		{
			name: "Redirect",
			url:  ts.URL + "/redirect",
			options: map[string]interface{}{
				"redirect": "follow",
			},
			expected: map[string]string{"message": "test response"},
		},
		{
			name: "Error response",
			url:  ts.URL + "/error",
			options: map[string]interface{}{
				"method": "GET",
			},
			expected: "error response",
		},
		{
			name: "Timeout",
			url:  ts.URL + "/timeout",
			options: map[string]interface{}{
				"method": "GET",
			},
			expected: "timeout response",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Execute the fetch request
			ctx.Eval(fmt.Sprintf(`
				fetch("%s", %s)
					.then(res => res.text())
					.then(text => {
						try {
							const json = JSON.parse(text);
							globalThis.result = json;
						} catch {
							globalThis.result = text;
						}
					})
					.catch(err => {
						globalThis.result = err;
					});
			`, tt.url, toJSON(tt.options)))

			// Wait for the response
			Loop(func() {
				result := ctx.GlobalObject().Get("result")
				if result != tt.expected {
					t.Errorf("expected %v, got %v", tt.expected, result)
				}
			})
		})
	}
}

func TestHeaders(t *testing.T) {
	Loop, ctx := core.Como("")
	defer ctx.Free()

	// Test Headers constructor
	code := `
		const headers = new Headers({
			'Content-Type': 'application/json',
			'X-Custom': 'value'
		});

		headers.append('X-Custom', 'value2');
		headers.set('X-New', 'new-value');

		const result = {
			hasContentType: headers.has('Content-Type'),
			getContentType: headers.get('Content-Type'),
			getCustom: headers.get('X-Custom'),
			getNew: headers.get('X-New'),
			entries: Array.from(headers.entries()),
			keys: Array.from(headers.keys()),
			values: Array.from(headers.values())
		};

		globalThis.result = result;
	`

	ctx.Eval(code)

	// Wait for the response and verify headers
	Loop(func() {
		result := ctx.GlobalObject().Get("result").(js.Value)
		hasContentType := result.Get("hasContentType").(js.Value)
		if !hasContentType.IsBool() || hasContentType.String() != "true" {
			t.Error("Expected Content-Type header to exist")
		}
		if result.Get("getContentType").(js.Value).String() != "application/json" {
			t.Error("Expected Content-Type to be application/json")
		}
		if result.Get("getCustom").(js.Value).String() != "value, value2" {
			t.Error("Expected X-Custom to be value, value2")
		}
		if result.Get("getNew").(js.Value).String() != "new-value" {
			t.Error("Expected X-New to be new-value")
		}
	})
}

func TestFormData(t *testing.T) {
	Loop, ctx := core.Como("")
	defer ctx.Free()

	// Test FormData
	code := `
		const formData = new FormData();
		formData.append('text', 'value');
		formData.append('file', {
			name: 'test.txt',
			value: new Uint8Array([1, 2, 3])
		});

		const result = {
			contentType: formData.getHeaders()
		};

		globalThis.result = result;
	`

	ctx.Eval(code)

	// Wait for the response and verify FormData
	Loop(func() {
		result := ctx.GlobalObject().Get("result").(js.Value)
		if result.Get("contentType").(js.Value).String() == "" {
			t.Error("Expected FormData to have content type")
		}
	})
}

// Helper function to convert Go map to JSON string
func toJSON(v interface{}) string {
	if v == nil {
		return "undefined"
	}
	b, _ := json.Marshal(v)
	return string(b)
}

File: fetch.test.ts
import { suite, assert } from './mod';
import { createServer } from 'http';

const test = suite('fetch');

test.only('fetch requests', async () => {
    // Create a test server
    const server = createServer((req, res) => {
        switch (req.url) {
            case '/test':
                // res.setHeader('Content-Type', 'application/json');
                res.end(JSON.stringify({ message: 'test response' }));
                break;
            case '/redirect':
                res.writeHead(302, { Location: '/test' });
                res.end();
                break;
            case '/error':
                res.writeHead(500);
                res.end('error response');
                break;
            case '/timeout':
                setTimeout(() => {
                    res.end('timeout response');
                }, 2000);
                break;
            default:
                res.writeHead(404);
                res.end();
        }
    });

    await new Promise<void>((resolve) => {
        server.listen(8080, () => {
            resolve();
        });
    });

    const port = 8080;
    const baseUrl = `http://localhost:${port}`;

    try {
        // Test GET request
        const getResponse = await fetch(`${baseUrl}/test`, { method: 'GET' });
        const getData = await getResponse.json();
        assert.equal(getData, { message: 'test response' });

        // Test POST request
        const postResponse = await fetch(`${baseUrl}/test`, {
            method: 'POST',
            body: 'test body'
        });
        const postData = await postResponse.json();
        assert.equal(postData, { message: 'test response' });

        // Test redirect
        const redirectResponse = await fetch(`${baseUrl}/redirect`, {
            redirect: 'follow'
        });
        const redirectData = await redirectResponse.json();
        assert.equal(redirectData, { message: 'test response' });

        // Test error response
        const errorResponse = await fetch(`${baseUrl}/error`, { method: 'GET' });
        const errorText = await errorResponse.text();
        assert.equal(errorText, 'error response');

        // Test timeout
        const timeoutResponse = await fetch(`${baseUrl}/timeout`, { method: 'GET' });
        const timeoutText = await timeoutResponse.text();
        assert.equal(timeoutText, 'timeout response');
    } catch(e) {
        console.log(e)
    } finally {
        // server.close();
    }
});

test('headers', async () => {
    const headers = new Headers({
        'Content-Type': 'application/json',
        'X-Custom': 'value'
    });

    headers.append('X-Custom', 'value2');
    headers.set('X-New', 'new-value');

    assert.ok(headers.has('Content-Type'));
    assert.equal(headers.get('Content-Type'), 'application/json');
    assert.equal(headers.get('X-Custom'), 'value, value2');
    assert.equal(headers.get('X-New'), 'new-value');
});

test('form data', async () => {
    const formData = new FormData();
    formData.append('text', 'value');
    formData.append('file', new Blob([new Uint8Array([1, 2, 3])], { type: 'text/plain' }));
});

test.run()